pycbc/waveform/waveform_modes.py

```diff
@@ -0,0 +1,491 @@
+"""
+
+from string import Formatter
+import lal
+import lalsimulation
+
+from pycbc.types import (TimeSeries, FrequencySeries)
+from .waveform import (props, _check_lal_pars, check_args)
+from . import parameters
+
+
+def _formatdocstr(docstr):
+    """Utility for formatting docstrings with parameter information.
+    """
+    return docstr.format(
+        **{_p[1]: getattr(parameters, _p[1]).docstr(
+            prefix="    ", include_label=False).lstrip(' ')
+           for _p in Formatter().parse(docstr) if _p[1] is not None
+           })
+
+
+def _formatdocstrlist(docstr, paramlist, skip_params=None):
+    """Utility for formatting docstrings with parameter information.
+    """
+    if skip_params is None:
+        skip_params = []
+    pl = '\n'.join([_p.docstr(prefix="    ", include_label=False)
+                    for _p in paramlist if _p not in skip_params]).lstrip(' ')
+    return docstr.format(params=pl)
+
+
+def sum_modes(hlms, inclination, phi):
+    """Applies spherical harmonics and sums modes to produce a plus and cross
+    polarization.
+
+    Parameters
+    ----------
+    hlms : dict
+        Dictionary of ``(l, m)`` -> complex ``hlm``. The ``hlm`` may be a
+        complex number or array, or complex ``TimeSeries``. All modes in the
+        dictionary will be summed.
+    inclination : float
+        The inclination to use.
+    phi : float
+        The phase to use.
+
+    Returns
+    -------
+    complex float or array
+        The plus and cross polarization as a complex number. The real part
+        gives the plus, the negative imaginary part the cross.
+    """
+    out = None
+    for mode in hlms:
+        ell, m = mode
+        hlm = hlms[ell, m]
+        ylm = lal.SpinWeightedSphericalHarmonic(inclination, phi, -2, ell, m)
+        if out is None:
+            out = ylm * hlm
+        else:
+            out += ylm * hlm
+    return out
+
+
+def default_modes(approximant):
+    """Returns the default modes for the given approximant.
+    """
+    # FIXME: this should be replaced to a call to a lalsimulation function,
+    # whenever that's added
+    if approximant in ['IMRPhenomXPHM', 'IMRPhenomXHM']:
+        # according to arXiv:2004.06503
+        ma = [(2, 2), (2, 1), (3, 3), (3, 2), (4, 4)]
+        # add the -m modes
+        ma += [(ell, -m) for ell, m in ma]
+    elif approximant in ['IMRPhenomPv3HM', 'IMRPhenomHM']:
+        # according to arXiv:1911.06050
+        ma = [(2, 2), (2, 1), (3, 3), (3, 2), (4, 4), (4, 3)]
+        # add the -m modes
+        ma += [(ell, -m) for ell, m in ma]
+    elif approximant.startswith('NRSur7dq4'):
+        # according to arXiv:1905.09300
+        ma = [(ell, m) for ell in [2, 3, 4] for m in range(-ell, ell+1)]
+    else:
+        raise ValueError("I don't know what the default modes are for "
+                         "approximant {}, sorry!".format(approximant))
+    return ma
+
+
+def get_glm(ell, m, theta):
+    """The inclination part of the ylm."""
```

Comment:
I can see what this is doing, but the docstring is confusing. Am I right in saying that this will always return a real number, and then this is the magnitude of the Y_lm?

Reply:
Yes, that's right. I've fleshed out the doc string.